apps/dashr-tsne/app.R

```diff
@@ -0,0 +1,1195 @@
+      #return(htmlP(paste(glue('{clickDat[[1]][2]}'))))
+      # return(htmlP(paste(glue('{imageIndex}'))))
+      
+      
+      
+      imageVec<- as.numeric(datTbl[imageIndex, ])
+      #
+      
+      
+      if(dataset=='cifar_gray_3000'){ imageMtx <- matrix(imageVec, nrow=32)} else {imageMtx <- matrix(imageVec, nrow=28)}
+      DIGIT <- writePNG(t(imageMtx))
+      DIGIT_B64 <- base64encode(DIGIT)
+      
+      return(htmlImg(src =  glue('data:image/png;base64, ', DIGIT_B64), style =  l.(height='25vh', display='block', margin='auto')))
+      
+    } else{
+      return(l.())
+    }
+  }
+  
+)
+#########CALLBACKS for DEMO ends here
+#
+#
+#########CALLBACKS for custom data begins
+#
+
+#helper function
+parse_contents <- function(contents, filename){
+  contents_parsed <- stringr::str_split(contents, ',')
+  decoded <- jsonlite::base64_dec(contents_parsed[[1]][2]) %>% rawToChar(.)
+  if(str_detect(filename, '.csv')){
+    DT <- fread(decoded)
+    
+  } else if(str_detect(filename, '.xls')){
+    DT <- read_xls(decoded)
+  } else{
+    print("Invalid file")
+    return(list(NULL, message='INVALID FILE'))
+  }
+  
+  return(list(dt=DT, message=glue('Data {filename} has been successfully uploaded')))
+}
+
+
+# main callback
+# Store the uploaded data to the hidden data htmlDiv container
+# 
+app$callback(
+  
+  output = list(id='data-df-and-message', property='children')
+  ,
+  params = list(
+    
+    input(id='upload-data', property='contents')
+    ,
+    input(id='upload-data', property='filename')
+    
+  )
+  ,
+  function(contents, filename){
+    contents_parsed <- parse_contents(contents, filename)
+    if(is.na(filename)){return(NULL)}
+    #decoded <- base64enc::base64decode(contents_parsed[2])
+    if(is.na(contents_parsed)){return(NULL)}
+    dt <-contents_parsed$dt %>% toJSON(., force=TRUE)
+    message <- contents_parsed$message %>% toJSON(., force=TRUE)
+    
+    C <-c(dt, message) %>% toJSON(., force=TRUE)
+    return(C)
+  }
+)
+
+
+app$callback(
+  
+  output = list(id='upload-data-message', property='children')
+  ,
+  params = list(
+    
+    input(id='data-df-and-message', property='children')
+    
+    
+  )
+  ,
+  function(JSON){
+    if(is.na(JSON)){return("NO DATA LOADED")}
+    
+    box <- fromJSON(JSON)
+    
+    return(box[2] %>% fromJSON)
+  }
+  
+)
+
+
+app$callback(
+  
+  output = list(id='label-df-and-message', property='children')
+  ,
+  params = list(
+    
+    input(id='upload-label', property='contents')
+    ,
+    input(id='upload-label', property='filename')
+    
+  )
+  ,
+  function(contents, filename){
+    contents_parsed <- parse_contents(contents, filename)
+    if(is.na(filename)){return(NULL)}
+    #decoded <- base64enc::base64decode(contents_parsed[2])
+    if(is.na(contents_parsed)){return(NULL)}
+    dt <-contents_parsed$dt %>% toJSON(., force=TRUE)
+    message <- contents_parsed$message %>% toJSON(., force=TRUE)
+    
+    C <-c(dt, message) %>% toJSON(., force=TRUE)
```

Comment:
I could be wrong, but I think the `.,` is already implied with the pipe, you could remove it and just have 

```
C <- c(dt, message) %>%
  toJSON(force = TRUE)
```

Reply:
It is implied with the pipe. I like it to keep it there for readability. Should I remove them?

Reply:
It's not that important I guess... I keep them out, some people keep them in :man_shrugging: